dora/src/bytecode/generator.rs

```diff
@@ -488,6 +488,77 @@ impl<'a, 'ast> AstBytecodeGen<'a, 'ast> {
         }
     }
 
+    fn visit_expr_delegation(&mut self, expr: &ExprDelegationType, dest: DataDest) -> Register {
+        let call_type = self.src.map_calls.get(expr.id).unwrap().clone();
+        let fct_id = call_type.fct_id().unwrap();
+
+        let fct = self.vm.fcts.idx(fct_id);
+        let fct = fct.read();
+
+        let callee_id = if fct.kind.is_definition() {
+            unimplemented!()
+        } else {
+            fct_id
+        };
+
+        let callee = self.vm.fcts.idx(callee_id);
+        let callee = callee.read();
+
+        if let FctKind::Builtin(_intrinsic) = callee.kind {
+            unimplemented!()
+        }
+
+        let return_type = if dest.is_effect() {
+            BuiltinType::Unit
+        } else {
+            self.specialize_type_for_call(&call_type, callee.return_type)
+        };
+        let arg_types = callee
+            .params_with_self()
+            .iter()
+            .map(|&arg| self.specialize_type_for_call(&call_type, arg).into())
+            .collect::<Vec<BytecodeType>>();
+        let num_args = arg_types.len();
+
+        let return_reg = if return_type.is_unit() {
+            Register::invalid()
+        } else {
+            self.ensure_register(dest, return_type.into())
+        };
+
+        let start_reg = if num_args > 0 {
+            self.gen.add_register_chain(&arg_types)
+        } else {
+            Register::zero()
+        };
+
+        let arg_start_reg = if callee.has_self() {
+            let self_id = self.src.var_self().id;
+            let self_reg = self.var_reg(self_id);
+
+            self.gen.emit_mov_ptr(start_reg, self_reg);
+            start_reg.offset(1)
+        } else {
+            start_reg
+        };
+
+        for (idx, arg) in expr.args.iter().enumerate() {
+            let arg_reg = arg_start_reg.offset(idx);
+            self.visit_expr(arg, DataDest::Reg(arg_reg));
+        }
+
+        match *call_type {
+            CallType::Ctor(_, _) => {
+                self.gen
```

Comment:
I would add an `assert(return_type.is_unit())` here. Otherwise this code might silently generate wrong in the future.

Reply:
This means implicitly, that `ensure_register` a few lines above is incorrect.

Reply:
Why? AFAICS ensure_register can handle all types, only this code would silently generate the wrong bytecode when the result is actually needed.

Reply:
Oh wait, yes you are right. Constructors should always return void, so the `ensure_register` above is not needed here.